File: java/gradle/src/main/java/com/css/challenge/client/Action.java
package com.css.challenge.client;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

/** Action is a json-friendly representation of an action. */
public class Action {
  public static final String PLACE = "place";
  public static final String MOVE = "move";
  public static final String PICKUP = "pickup";
  public static final String DISCARD = "discard";

  public static final String HEATER = "heater";
  public static final String COOLER = "cooler";
  public static final String SHELF = "shelf";

  private final long timestamp; // unix timestamp in microseconds
  private final String id; // order id
  private final String action; // place, move, pickup or discard
  private final String target; // heater, cooler or shelf. Target is the destination for move

  public Action(Instant timestamp, String id, String action, String target) {
    this.timestamp = ChronoUnit.MICROS.between(Instant.EPOCH, timestamp);
    this.id = id;
    this.action = action;
    this.target = target;
  }

  public long getTimestamp() {
    return timestamp;
  }

  public String getId() {
    return id;
  }

  public String getAction() {
    return action;
  }

  public String getTarget() {
    return target;
  }

  @Override
  public String toString() {
    return "{timestamp: "
        + timestamp
        + ", id: "
        + id
        + ", action: "
        + action
        + ", target: "
        + target
        + "}";
  }
}

File: kotlin/gradle/src/main/kotlin/client/Action.kt
package com.css.challenge.client

import kotlin.time.Instant
import kotlinx.serialization.Serializable

const val PLACE = "place"
const val PICKUP = "pickup"
const val MOVE = "move"
const val DISCARD = "discard"

const val HEATER = "heater"
const val COOLER = "cooler"
const val SHELF = "shelf"

@Serializable
data class Action(
    val timestamp: Long, // unix timestamp in microseconds
    val id: String,      // order id
    val action: String,  // place, move, pickup or discard
    val target: String,  // heater, cooler or shelf. Target is the destination for move
) {
    constructor(timestamp: Instant, id: String, action: String, target: String) : this(timestamp.toEpochMilliseconds() * 1000L, id, action, target)
}
